### ISSUE TITLE
Typescript typing do not match the example with other extensions in README

### ISSUE THREAD

**Issue by thupi** [dump247/storybook-state#8]
Hi :-)

This is an awesome addon for Storybook.

However i have a hard time getting this to work with storybook info due to the typings. It seams to me like the typings do not match the usage example in the README :-)

**Comment by dump247** [comment]
I apologize for that. I am not familiar with typescript. The typescript definition file was contributed by someone else and I forgot to update it.

Does this look correct? This is based on my trial and error with the [typescript playground](https://www.typescriptlang.org/play/), so I am not sure if it works or not.
```typescript
import { Renderable, RenderFunction } from '@storybook/react';

export type Store<T> = {
  /**
   * @description
   * Object that contains the current state.
   */
  state: T;

  /**
   * @description
   * Set the given state keys. The state parameter is an object with the keys and values to set.
   * This only sets/overwrites the specific keys provided.
   */
  set(nextState: Partial<T>): void;

  /**
   * @description
   * Reset the store to the initial state.
   */
  reset(): void;
};

export type StateRender<T> = ({ store }: { store: Store<T> }) => Renderable | Renderable[];
export type StoryRender<T> = (cb: StateRender<T>) => Renderable | Renderable[]
export function withState<T extends {}>(state: T): StoryRender<T>;
```

**Comment by thupi** [comment]
Hi @dump247 :-) 

I tried your typing and while it looks right to me i was not able to get it working :-) 

I spent a bit time on adjusting the types and so on i hit fundamental issue. The @types/storybook__react do not allow any props in the type called RenderFunction. @storybook/addon-info uses that type which means their typing wont allow for props within the withInfo function. :-) 

I ended up with these two type errors:
```
ERROR in ./src/Modal/ActionModal/ActionModal.stories.tsx
[tsl] ERROR in /opt/wiseflow/flow-ui/src/Modal/ActionModal/ActionModal.stories.tsx(17,3)
      TS2345: Argument of type 'ComponentClass<{}> | StatelessComponent<{}> | Element | Renderable[]' is not assignable to parameter of type 'RenderFunction'.
  Type 'ComponentClass<{}>' is not assignable to type 'RenderFunction'.
    Type 'ComponentClass<{}>' provides no match for the signature '(): ComponentClass<{}> | StatelessComponent<{}> | Element | Renderable[]'.

ERROR in ./src/Modal/ActionModal/ActionModal.stories.tsx
[tsl] ERROR in /opt/wiseflow/flow-ui/src/Modal/ActionModal/ActionModal.stories.tsx(18,32)
      TS2345: Argument of type '({ store }: any) => Element' is not assignable to parameter of type 'RenderFunction'.
```

With my story looking like the this:

```
import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, text, boolean, number } from '@storybook/addon-knobs/react';
import { withState } from '@dump247/storybook-state';

// Storybook
const stories = storiesOf('Modal', module);
stories.addDecorator(withKnobs);

// Components
import ActionModal from './ActionModal';

// Stories
stories.add(
  'ActionModal',
  withState({ initialState: '' })(
    withInfo(`Some cool info`)(({ store }: any) => (
      <div>
        <ActionModal open={true}>
          {store.state.checked ? 'Open' : 'False'}
        </ActionModal>
        <button onClick={() => store.set({ checked: !store.state.checked })}>
          Toggle
        </button>
      </div>
    ))
  )
);
```

I got it working with this typing:
```
import { Renderable, RenderFunction } from '@storybook/react';
export type Store<T> = {
  /**
   * @description
   * Object that contains the current state.
   */
  state: T;

  /**
   * @description
   * Set the given state keys. The state parameter is an object with the keys and values to set.
   * This only sets/overwrites the specific keys provided.
   */
  set(nextState: Partial<T>): void;

  /**
   * @description
   * Reset the store to the initial state.
   */
  reset(): void;
};

export type StorybookStateCallback<T> = (store: Store<T>) => Renderable | Renderable[];
export function withState<T extends {}>(state: T, callback?: StorybookStateCallback<T>): any // RenderFunction;
```
And i had to adjust the @types/storybook__react's RenderFunction type from `type RenderFunction = () => Renderable | Renderable[];` to `type RenderFunction = (props: any) => Renderable | Renderable[];` in order to let the store be passed into the RenderFunction :-)

I used the same Story as mentiond above :-) Hope that helps.

I dont have the time to investigate further at the moment. :-)

**Comment by dump247** [comment]
Should be resolved in [v1.5.0](4de3aeca728d2def7c80ec284ea1ea9aea87e258). I used the TS definition @thupi provided.

**Comment by thupi** [comment]
@dump247 That should not work :-) ? The Typings at @types/storybook__react do not allow for props inside the RenderFunction as i mentioned before :-) 

In order to get the typing work one would have to change the types from storybook-react :-) ?

Maybe this would be a topic that should be at the storybook-react repo :-) I'll try to mention this issue at storybook's repo :-) 

The updated typing would work if or when storybook changes the typing at their repo :-)

**Comment by dump247** [comment]
Ok, I understand now. I misunderstood your previous comment. Reopening since this is still an issue.

**Comment by PaulieScanlon** [comment]
i love withState it's great when developing 'dumb components' ... i love typescript to band this is a real bummer. Any update on if it'll be fixed? i'd offer to do it but not sure i've got the skills :-)

**Comment by zhenwenc** [comment]
This library is great, but it would be a little bit better if its also type safe :-) 

Inspired by this library, here is what I solve the issue with a wrapper component, hope it helps a little:
```
export interface WithStateProps<S> {
  initialState: S;

  children: (
    state: S,
    api: {
      setState: <K extends keyof S>(
        state: Pick<S, K> | S | null,
        callback?: () => void
      ) => void;
    }
  ) => React.ReactNode;
}

export class WithState<S = {}> extends React.Component<WithStateProps<S>, S> {
  readonly state: S = this.props.initialState;

  render() {
    const { children } = this.props;
    return children(this.state, {
      setState: (state, cb) => this.setState(state, cb),
    });
  }
}
```

How to use:

```
<WithState initialState={{ counter: 0 }}>
  {({ counter }, { setState }) => (
    <button onClick={() => setState({ counter: counter + 1 })}>
      count: {counter}
    </button>
  )}
</WithState>
``` 

The state type is inferred from `initialState`, or you could manually declare the shape by:

```
<WithState<{ counter: number }> initialState={{ counter: 0 }}>
  ...
</WithState>
```
Note: this may require TS 2.9 (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-9.html#generic-type-arguments-in-jsx-elements)

**Comment by dump247** [comment]
I think this has been resolved in #11. Please reopen if you still have issues.
